Maskbook/Wallet/WalletService/NFTScan/NFTScanProvider.swift

```diff
@@ -0,0 +1,338 @@
+        cursor: String,
+        limit: Int
+    ) -> AnyPublisher<NFTScanAssetDataModel, Error>? {
+        guard let address = maskUserDefaults.defaultAccountAddress else {
+            return nil
+        }
+
+        let decoder = JSONDecoder()
+        var fetchTokenURLComponents = URLComponents(
+            url: baseURL.appendingPathComponent("account/own/\(address)"),
+            resolvingAgainstBaseURL: false
+        )
+
+        fetchTokenURLComponents?.queryItems = [
+            URLQueryItem(name: "erc_type", value: "erc721"),
+            URLQueryItem(name: "limit", value: "\(limit)"),
+            URLQueryItem(name: "cursor", value: "\(cursor)"),
+        ]
+
+        guard let requestURL = fetchTokenURLComponents?.url else {
+            return nil
+        }
+
+        var fetchTokenRequest = URLRequest(url: requestURL)
+        fetchTokenRequest.setValue(APIKey.NFTSCAN, forHTTPHeaderField: "x-api-key")
+        let tokenPublisher =
+            session.dataTaskPublisher(for: fetchTokenRequest)
+                .tryMap { element -> Data in
+                    guard let httpResponse = element.response as? HTTPURLResponse,
+                          httpResponse.statusCode == 200 else {
+                        throw URLError(.badServerResponse)
+                    }
+                    let json = try JSONSerialization.jsonObject(
+                        with: element.data,
+                        options: []
+                    ) as? [String: Any]
+                    if let assets = json?["data"] {
+                        return try JSONSerialization.data(withJSONObject: assets, options: [])
+                    }
+                    return element.data
+                }
+                .decode(type: NFTScanAssetDataModel.self, decoder: decoder)
+                .map {
+                    $0
+                }
+                .eraseToAnyPublisher()
+
+        return tokenPublisher
+    }
+
+    private func retrieveTransactionsByAddress(
+        contractAddress: String,
+        tokenID: String
+    ) -> AnyPublisher<NFTScanTransactionDataModel, Error>? {
+        let decoder = JSONDecoder()
+        var fetchTokenURLComponents = URLComponents(
+            url: baseURL.appendingPathComponent("transactions/\(contractAddress)/\(tokenID)"),
+            resolvingAgainstBaseURL: false
+        )
+
+        fetchTokenURLComponents?.queryItems = [
+            URLQueryItem(name: "limit", value: "100"),
+        ]
+
+        guard let requestURL = fetchTokenURLComponents?.url else {
+            return nil
+        }
+
+        var fetchTokenRequest = URLRequest(url: requestURL)
+        fetchTokenRequest.setValue(APIKey.NFTSCAN, forHTTPHeaderField: "x-api-key")
+        let tokenPublisher =
+            session.dataTaskPublisher(for: fetchTokenRequest)
+                .tryMap { element -> Data in
+                    guard let httpResponse = element.response as? HTTPURLResponse,
```

Comment:
We should extract this mapper logic into a static function and replace the duplicated code here and below.